### PR TITLE
KAFKA-4377: remove deprecated scala.collection.JavaConversions calls

### PR DESCRIPTION
**Pull request by leachbj**
JavaConversions are deprecated in 2.12 in favour of JavaConverters.

### CODE REVIEW

File: core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala [hunk 1]
```diff
@@ -696,9 +696,9 @@ private[kafka] class ZookeeperConsumerConnector(val config: ConsumerConfig,
             if (topicRegistry.size == 0)
               new java.util.HashMap[String, java.util.Set[java.lang.Integer]]
             else
-              mapAsJavaMap(topicRegistry.map(topics =>
-                topics._1 -> topics._2.keys
-              ).toMap).asInstanceOf[java.util.Map[String, java.util.Set[java.lang.Integer]]]
+              topicRegistry.map(topics =>
+                topics._1 -> topics._2.keys.asInstanceOf[mutable.Set[Integer]].asJava
```
Comment (leachbj) [inline core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala]: this is actually a bug fix to KAFKA-2284 and is the cause for the test failure for this PR as the test incorrectly verifies the resut is a scala `Set`

I have separated out the bugfix into https://github.com/apache/kafka/pull/2102 so that PR will need to be merged first.

File: core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala [hunk 2]
```diff
@@ -27,6 +27,9 @@ import org.apache.kafka.common.protocol.ApiKeys
 import org.junit.Assert._
 import org.junit.{After, Before, Test}
 
+import scala.collection.JavaConverters._
+import scala.collection.mutable
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala]: Why is this required?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 3]
```diff
@@ -757,18 +757,14 @@ object KafkaConfig {
...
 
   /**
     * Check that property names are valid
     */
   def validateNames(props: Properties) {
-    import scala.collection.JavaConversions._
-    val names = configDef.names()
-    for (name <- props.keys)
+    val names = configDef.names().asScala
+    for (name <- props.keys.asScala.asInstanceOf[Set[String]])
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:767]: Can we not avoid the cast?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 4]
```diff
@@ -757,18 +757,14 @@ object KafkaConfig {
 
   }
 
-  def configNames() = {
-    import scala.collection.JavaConversions._
-    configDef.names().toList.sorted
-  }
+  def configNames() = configDef.names().asScala.toList.sorted
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:760]: I don't think we need `toList` here

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 5]
```diff
@@ -1031,11 +1029,11 @@ class KafkaApis(val requestChannel: RequestChannel,
         JoinGroupResponse.UNKNOWN_PROTOCOL,
         JoinGroupResponse.UNKNOWN_MEMBER_ID, // memberId
         JoinGroupResponse.UNKNOWN_MEMBER_ID, // leaderId
-        Map.empty[String, ByteBuffer])
+        Map.empty[String, ByteBuffer].asJava)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: It seems like you used `Collections.empty*` methods in a few places. Should we use `emptyMap` here then?

File: core/src/main/scala/kafka/controller/ReplicaStateMachine.scala [hunk 6]
```diff
@@ -300,7 +300,7 @@ class ReplicaStateMachine(controller: KafkaController) extends Logging {
   }
 
   def replicasInDeletionStates(topic: String): Set[PartitionAndReplica] = {
-    val deletionStates = Set(ReplicaDeletionStarted, ReplicaDeletionSuccessful, ReplicaDeletionIneligible)
+    val deletionStates = Set[ReplicaState](ReplicaDeletionStarted, ReplicaDeletionSuccessful, ReplicaDeletionIneligible)
```
Comment (ijuma) [inline core/src/main/scala/kafka/controller/ReplicaStateMachine.scala:303]: Why do we need the type parameter here?

File: core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala (hunk 1 above)
Comment (ijuma) [reply core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala]: It would be good if this PR did not change behaviour.

File: core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala [hunk 7]
```diff
@@ -755,14 +755,13 @@ private[kafka] class ZookeeperConsumerConnector(val config: ConsumerConfig,
                 case (topic, partitionOwnerShips) =>
                   val partitionOwnershipForTopicScalaMap = partitionOwnerShips.map({
                     case (topicAndPartition, consumerThreadId) =>
-                      topicAndPartition.partition -> consumerThreadId
-                  })
-                  topic -> mapAsJavaMap(collection.mutable.Map(partitionOwnershipForTopicScalaMap.toSeq:_*))
-                    .asInstanceOf[java.util.Map[java.lang.Integer, ConsumerThreadId]]
+                      topicAndPartition.partition.asInstanceOf[Integer] -> consumerThreadId
```
Comment (ijuma) [inline core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala]: Instead of `asInstanceOf`, we could use `: Integer`, right?

File: core/src/main/scala/kafka/api/FetchResponse.scala [hunk 8]
```diff
@@ -138,7 +138,7 @@ class TopicDataSend(val dest: String, val topicData: TopicData) extends Send {
   buffer.rewind()
 
   private val sends = new MultiSend(dest,
-                            JavaConversions.seqAsJavaList(topicData.partitionData.toList.map(d => new PartitionDataSend(d._1, d._2))))
+    JavaConverters.seqAsJavaListConverter[Send](topicData.partitionData.map(d => new PartitionDataSend(d._1, d._2))).asJava)
```
Comment (ijuma) [inline core/src/main/scala/kafka/api/FetchResponse.scala]: This is a bit annoying. Is that the most concise non-deprecated option available?

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, looks good overall, left a few minor comments.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 4 above)
Comment (leachbj) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:760]: `configDef.names()` gives a Set, before the change there was an implicit conversion on the result of names() to convert it to a Set, now its just explicit.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 3 above)
Comment (leachbj) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:767]: Unfortuantely props is a `Properties` == `Hashtable<Object, Object>` so the `props.keys` ends up as an `AnyRef`.

File: core/src/main/scala/kafka/controller/ReplicaStateMachine.scala (hunk 6 above)
Comment (leachbj) [reply core/src/main/scala/kafka/controller/ReplicaStateMachine.scala:303]: The values are all objects so they end up as `ReplicaState with Product with Serializable` so the `contains(r._2)` fails to typecheck.

File: core/src/main/scala/kafka/api/FetchResponse.scala (hunk 8 above)
Comment (leachbj) [reply core/src/main/scala/kafka/api/FetchResponse.scala]: Its the one that will be used but this works and reads better;

`topicData.partitionData.map(d => new PartitionDataSend(d._1, d._2) : Send).asJava)`

File: core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala [hunk 9]
```diff
@@ -696,9 +696,9 @@ private[kafka] class ZookeeperConsumerConnector(val config: ConsumerConfig,
             if (topicRegistry.size == 0)
               new java.util.HashMap[String, java.util.Set[java.lang.Integer]]
             else
-              mapAsJavaMap(topicRegistry.map(topics =>
-                topics._1 -> topics._2.keys
-              ).toMap).asInstanceOf[java.util.Map[String, java.util.Set[java.lang.Integer]]]
+              topicRegistry.map(topics =>
+                topics._1 -> topics._2.keys   // note this is incorrect, see KAFKA-2284
+              ).toMap.asJava.asInstanceOf[java.util.Map[String, java.util.Set[java.lang.Integer]]]
```
Comment (leachbj) [inline core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala:701]: reverted to buggy behaviour now

File: core/src/main/scala/kafka/admin/AdminClient.scala [hunk 10]
```diff
@@ -76,11 +74,11 @@ class AdminClient(val time: Time,
   def listGroups(node: Node): List[GroupOverview] = {
     val response = send(node, ApiKeys.LIST_GROUPS, new ListGroupsRequest()).asInstanceOf[ListGroupsResponse]
     Errors.forCode(response.errorCode()).maybeThrow()
-    response.groups().map(group => GroupOverview(group.groupId(), group.protocolType())).toList
+    response.groups().asScala.map(group => GroupOverview(group.groupId(), group.protocolType())).toList
   }
 
   private def findAllBrokers(): List[Node] = {
-    val request = new MetadataRequest(List[String]())
+    val request = new MetadataRequest(Collections.emptyList[String])
```
Comment (leachbj) [inline core/src/main/scala/kafka/admin/AdminClient.scala:81]: Now just uses the java APIs where possible rather than converting scala instances

File: core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala [hunk 11]
```diff
@@ -27,6 +27,8 @@ import org.apache.kafka.common.protocol.ApiKeys
 import org.junit.Assert._
 import org.junit.{After, Before, Test}
 
+import scala.collection.JavaConverters._
```
Comment (leachbj) [inline core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala:30]: revert

**Review by leachbj — COMMENTED** [review]
Just the one unused import to sort out.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 3 above)
Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:767]: It turns out that this method is not used, so will remove it before merging.

**Comment by ijuma** [comment]
Thanks for the PR, LGTM. I merged it to trunk after a few minor tweaks.
